compiler/rustc_const_eval/src/interpret/eval_context.rs

```diff
@@ -1141,18 +1139,20 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         self.raw_const_to_mplace(val)
     }
 
-    pub fn eval_mir_constant(
+    /// Evaluate a constant from the current frame.
+    /// This function relies on the fact that when the frame got pushed, all `required_consts` have been checked.
+    pub(super) fn eval_frame_mir_constant(
         &self,
-        val: &mir::Const<'tcx>,
+        const_: &mir::Const<'tcx>,
         span: Option<Span>,
         layout: Option<TyAndLayout<'tcx>>,
     ) -> InterpResult<'tcx, OpTy<'tcx, M::Provenance>> {
-        M::eval_mir_constant(self, *val, span, layout, |ecx, val, span, layout| {
-            let const_val = val.eval(*ecx.tcx, ecx.param_env, span).map_err(|err| {
-                // FIXME: somehow this is reachable even when POST_MONO_CHECKS is on.
-                // Are we not always populating `required_consts`?
```

Comment:
I finally solved this mystery. :) It was promoteds containing failing consts without recording them in their own `required_consts`.